scripts/continuous/prepare.sh

```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+target=${1//\//__}
+
+x86__test__net() {
+	make confload-x86/test/ping-target
+	make -j $(nproc)
+	cp build/base/bin/embox ./ping-target
+}
+
+if ! [ $target ] || ! type -t $target > /dev/null; then
+	echo nothing to prepare for \"$1\"
+	exit 0
+fi
+$target "$@"
```

Comment:
What is the point to introduce such a kind of preparation? Could it be executed just in place before ping test in continuous/net/run.sh?

There are already two stages - `build` + `run`. Now you introduce one additional stage and there are `prepare` + `build` + `run`. What is a strict division into `prepare` and others?  At the same time, x86/tests/fs modify start_script and rebuilds embox to test various FS types. So there is no even strict division into `build` and `run` since build `actually` happens during `run`. To keep things simple I'd even propose to combine `build` and `run` stages into a single one - `test`, which can modify modules options, start_script, build whatever required and so on, or just build and run qemu.

Another way is to prepare all the required things before `run`, while `run` builds nothings, justs perform testing. But in that case two stages are enough - `prepare` + `run`.

Reply:
Tests for `x86/test/net` change template as well. 
Without `prepare`, we would build `x86/test/net`, then `ping-target`, then `x86/test/net'`s (the template with changes). This would give us 3 clean builds, as each goes after `confload`
With `prepare`, we do `ping-target`, `x86/test/net`, `x86/test/net'`. It is 2 clean builds.

`prepare` are steps done before actual build

Build in `build` and build in `run` are different by nature. Build in `build` may fail, it will be a source code issue. Build in `test` fails less likely due to minor changes in config and no changes in code; if the build fails it's tests issue.

prepare + build + run is the semantically the one test procedure. 
Having multiple steps:
1. Allows quick overlook for failed step, you don't need to search for error in a big combined log (tests continue even after failed one) 
2. Each step have separate log, at least Actions interface is more responsive when step is unfolded and a small log appears.

Could you please elaborate simplicity of combined steps. What are its practical advantages?

Reply:
My point of view here is quite simple - it's better do not divide the things into multiple stages while we are not sure which stages are actually required.

A simple example, you add `prepare` stage. Should we move also `nfsd start` and friends there? Because it's also a kind of preparation that doesn't depend on embox build.

So if we claim that `prepare` means a sort of preparation that doesn't involve "target under test" (embox) at all - building some stuff under linux, starting services, etc., it would probably make sense. And only after we will decide so we can introduce a new stage. Probably you already has this in mind, that's why I started this discussion.

Reply:
By the way, if we would claim to start linux services in `prepare` then you should ping-target here too, not in `run` stage.

Reply:
> doesn't depend on embox build
is not the same
> prepare are steps done before actual build

I should stress that more: steps that are _required_ to run before the build. To build the ping-target before is required (to save one clean build), while run it is not. 

Ping-target cannot be started before the test. Actions are hang when leftover background job is running. I tried :)

> My point of view here is quite simple - it's better do not divide the things into multiple stages while we are not sure which stages are actually required.
> And only after that, we can decide whether to introduce a new stage. Probably you already have this in mind, that's why I started this discussion.

I'm confused by trying to save number of steps, that is, shell script invocations. Why is it important?
I had nothing in mind for optimizing number of steps. I care only about time required for tests to finish. This was a major point of pain, not number of steps
There are strong arguments for `prepare`: it saves one clean build (~1-2 min) and makes using Actions interface more pleasant.
Could you highlight, why one combined step is better that splitted?

Reply:
> This was a major point of pain, not number of steps
There are strong arguments for prepare: it saves one clean build (~1-2 min) and makes using Actions interface more pleasant.

How this prepare.sh saves time? If we would move it at `run` stage testing time will be longer?

Reply:
> Tests for `x86/test/net` change template as well.
> Without `prepare`, we would build `x86/test/net`, then `ping-target`, then `x86/test/net'`s (the template with changes). This would give us 3 clean builds, as each goes after `confload`
> With `prepare`, we do `ping-target`, `x86/test/net`, `x86/test/net'`. It is 2 clean builds.

Reply:
Ugh, now I understand what you are talking about It's a sort of optimization because `ping-target` and `x86/test/net` use different configs, so it is required to go through all these mybuild stages, right?

Reply:
An you optimized this in such a way that`modified x86/test/net' follow by `x86/test/net`?

Reply:
Yes, exactly

Reply:
But now I am wondering why you do not use the prepared `mk/.cache` overall in scripts/continues when some embox build is required instead of introducing new global `prepare` stage and keeping the builds order in mind?

Reply:
Ok, it is probably because of Travis didn't support this cache,

Reply:
Well, you keep all re-builds of the same template to follow each other, ok then.

Reply:
> But now I am wondering why you do not use the prepared `mk/.cache` overall in scripts/continues when some embox build/rebuild is required instead of introducing new global `prepare` stage and keeping the builds order in mind?
> Ok, it is probably because of Travis didn't support this cache,

I didn't get this at all.

> why don't you use the prepared `mk/.cache`  in scripts/continues when some embox build/rebuild is required 

I do. Mybuild does this automatically. mk/.cache contains Mybuild parsing results. mk/.cache is preserved between steps. But not between jobs

> introducing new global `prepare` stage and keeping the builds order in mind?

With prepare we are saving compiler invocations, `.o` files. They are not related to mk-cache

> Well, you keep all re-builds of the same template to follow each other, ok then.

That's correct

Reply:
> With prepare we are saving compiler invocations, .o files. They are not related to mk-cache

Which `.o` files? How they are used then?

Reply:
Ah, you mean files of `x86/test/net`, ok

Reply:
```
build: 
  make confload-x86/test/net
  make
test:
  sed -i '/smth/htms/' conf/mods.config
  make # incremental build, do reuse most of .o files
```

Reply:
Yes, thanks, I already said this one comment above https://github.com/embox/embox/pull/1801#discussion_r389894948